File: Array and Hashing/python/threeSum.py
# Problem Link: https://www.codingninjas.com/codestudio/problems/three-sum_6922132?utm_source=striver&utm_medium=website&utm_campaign=a_zcoursetuf


def triplet(n, nums):
    # sort the array
    nums.sort()
    # make a result list
    res = []

    for i in range(len(nums) - 2):
        if i == 0 or (i > 0 and nums[i] != nums[i - 1]):
            low, high = i + 1, len(nums) - 1
            target = -nums[i]
            while low < high:
                if nums[low] + nums[high] == target:
                    res.append([nums[i], nums[low], nums[high]])
                    while low < high and nums[low] == nums[low + 1]:
                        low += 1
                    while low < high and nums[high] == nums[high - 1]:
                        high -= 1
                    low += 1
                    high -= 1
                elif nums[low] + nums[high] < target:
                    low += 1
                else:
                    high -= 1

    return res


if __name__ == "__main__":
    nums = [-1, 0, 1, 2, -1, -4]
    n = len(nums)
    ans = triplet(n, nums)
    print("Triplets that sum to zero:")
    for t in ans:
        print(t)

File: Array and Hashing/cpp/threeSum.cpp
// Problem Link: https://www.codingninjas.com/codestudio/problems/three-sum_6922132?utm_source=striver&utm_medium=website&utm_campaign=a_zcoursetuf

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<vector<int>> triplet(int n, vector<int>& nums) {
    // sort the array
    sort(nums.begin(), nums.end());
    // make a result list
    vector<vector<int>> res;

    for (int i = 0; i < (int)nums.size() - 2; i++) {
        if (i == 0 || (i > 0 && nums[i] != nums[i - 1])) {
            int low = i + 1, high = (int)nums.size() - 1;
            int sum = 0 - nums[i];
            while (low < high) {
                if (nums[low] + nums[high] == sum) {
                    res.push_back({nums[i], nums[low], nums[high]});
                    while (low < high && nums[low] == nums[low + 1]) low++;
                    while (low < high && nums[high] == nums[high - 1]) high--;
                    low++;
                    high--;
                } else if (nums[low] + nums[high] < sum) {
                    low++;
                } else {
                    high--;
                }
            }
        }
    }
    return res;
}

int main() {
    vector<int> nums = {-1, 0, 1, 2, -1, -4};
    int n = nums.size();
    vector<vector<int>> ans = triplet(n, nums);
    cout << "Triplets that sum to zero:" << endl;
    for (auto& triplet : ans) {
        cout << "[" << triplet[0] << ", " << triplet[1] << ", " << triplet[2] << "]" << endl;
    }
    return 0;
}
